core/domain/classifier_services.py

```diff
@@ -127,6 +127,75 @@ def classify_string_classifier_rule(state, normalized_answer):
     return None
 
 
+def handle_classifier_training_job_creation(exploration, state_names):
+    """Creates ClassifierTrainingJobModel instances for all the state names
+    passed into the function.
+
+    Args:
+        exploration: Exploration. The Exploration domain object.
+        state_names: list(str). List of state names.
+    """
+    job_dicts_list = []
+    exp_id = exploration.id
+    exp_version = exploration.version
+    states = exploration.states
+    for state_name in state_names:
+        state = states[state_name]
+        training_data = state.get_training_data()
+        interaction_id = state.interaction.id
+        algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_id']
+        job_dicts_list.append({
+            'algorithm_id': algorithm_id,
+            'interaction_id': interaction_id,
+            'exp_id': exp_id,
+            'exp_version': exp_version,
+            'state_name': state_name,
+            'training_data': training_data,
+            'status': feconf.TRAINING_JOB_STATUS_NEW
+        })
+    create_multi_classifier_training_jobs(job_dicts_list)
+
+
+def handle_mappings_creation(exploration, state_names, change_list):
+    """Creates new TrainingJobExplorationMappingModel instances for all the
+    state names passed into the function. The mapping is created from the
+    new state to the ClassifierTrainingJob of the old state.
+
+    Args:
+        exploration: Exploration. The Exploration domain object.
+        state_names: list(str). List of state names.
+        change_list: list(dict). A list of changes introduced in this commit.
+    """
+    exp_id = exploration.id
+    current_exp_version = exploration.version
+    old_exp_version = current_exp_version - 1
+    for current_state_name in state_names:
+        old_state_name = current_state_name
+        classifier_training_job = get_classifier_training_job(
+            exp_id, old_exp_version, old_state_name)
+        # Check for renamed states.
+        if not classifier_training_job:
+            renamed_cmds = []
+            for change_dict in change_list:
```

Comment:
Why not filter this list outside of entire for loop and every time you need renamed cmds then use copy of original renamed cmds dict? That way you can avoid this and get better time complexity.